extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -85,3 +88,343 @@ export class ISignatureVerifier {
+     *    null, then no key with that keypair ID could be found even after
+     *    cachebusting. If it's a `Promise` that resolves to `null`, then key
+     *    data for that keypair ID was found but could not be imported
+     *    successfully; this most likely indicates signing service misbehavior.
+     *    The success case is a `Promise` that resolves to a `CryptoKey`.
+     *
+     * @private @const {?Object<string, {promise: !Promise<boolean>, keys: !Object<string, ?Promise<?webCrypto.CryptoKey>>}>}
+     */
+    this.signers_ = Services.cryptoFor(win).isPkcsAvailable() ? {} : null;
+
+    /**
+     * Gets a notion of current time, in ms.  The value is not necessarily
+     * absolute, so should be used only for computing deltas.  When available,
+     * the performance system will be used; otherwise Date.now() will be
+     * returned.
+     *
+     * @private @const {function(): number}
+     */
+    this.getNow_ = (win.performance && win.performance.now) ?
+        win.performance.now.bind(win.performance) : Date.now;
+  }
+
+  /**
+   * Fetches and imports the public keyset for the named signing service,
+   * without any cachebusting. Hopefully, this will hit cache in many cases
+   * and not make an actual network round-trip. This method should be called
+   * as early as possible, once it's known which signing service is likely to
+   * be used, so that the network request and key imports can execute in
+   * parallel with other operations.
+   *
+   * @param {string} signingServiceName
+   */
+  loadKeyset(signingServiceName) {
+    if (this.signers_ && !this.signers_[signingServiceName]) {
+      const keys = {};
+      const promise = this.fetchAndAddKeys_(keys, signingServiceName, null);
+      this.signers_[signingServiceName] = {promise, keys};
+    }
+  }
+
+  /**
+   * Extracts a cryptographic signature from `headers` and attempts to verify
+   * that it's the correct cryptographic signature for `creative`.
+   *
+   * As a precondition, `loadKeyset(signingServiceName)` must have already been
+   * called.
+   *
+   * @param {!ArrayBuffer} creative
+   * @param {!Headers} headers
+   * @param {function(string, !Object)} lifecycleCallback called for each AMP
+   *     lifecycle event triggered during verification
+   * @return {!Promise<!VerificationStatus>}
+   */
+  verify(creative, headers, lifecycleCallback) {
+    const signatureHeader = 'AMP-Fast-Fetch-Signature';
+    const signatureFormat =
+        /^([A-Za-z0-9._-]+):([A-Za-z0-9._-]+):([A-Za-z0-9+/]{341}[AQgw]==)$/;
+    if (!headers.has(signatureHeader)) {
+      return Promise.resolve(VerificationStatus.UNVERIFIED);
+    }
+    const headerValue = headers.get(signatureHeader);
+    const match = signatureFormat.exec(headerValue);
+    if (!match) {
+      // TODO(@taymonbeal, #9274): replace this with real error reporting
+      user().error('AMP-A4A', `Invalid signature header: ${headerValue}`);
```

Comment:
Do we really need to send the entire header value?  Seems unnecessary... how about just the signing service name?